# SHORTEST NEURAL NETWORK CODE. (FOR BEGINNERS)
# ADDED TENSORBOARD VISUALIZATION.

All you need is one library 'tensorflow'. Ofcourse we are using Keras(Highlevel API). It is now officially included in tesorflow

In [1]:
import tensorflow as tf

C:\Users\HP.DESKTOP-UCJPQBV\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


Load the MNIST dataset. The data has train and test splits with the features and labels separated.

In [2]:
(tr_x,tr_y),(ts_x,ts_y) = tf.keras.datasets.mnist.load_data()

In [3]:
print(tr_x.shape , tr_y.shape , ts_x.shape, ts_y.shape)

(60000, 28, 28) (60000,) (10000, 28, 28) (10000,)


Examine the first image in train set.

The input values are in range O-255(0 being pure white, 255 dark black). 
So scale them to 0-1 for faster calculation. (Advised, not mandatory).

Execute Print(tr_x[0]) to know about this.

In [4]:
tr_x = tr_x/255
ts_x = ts_x/255

Since, NN's need data to be flattened. Reshape the data.

In [5]:
import numpy as np
flat_tr_x = np.reshape(tr_x,[60000,784])
flat_ts_x = np.reshape(ts_x,[10000,784])

In [6]:
print(flat_tr_x.shape, flat_ts_x.shape)

(60000, 784) (10000, 784)


# TENSORBOARD
tb\demo is the file it is going to create in your default directory.

In [7]:
tb = tf.keras.callbacks.TensorBoard(log_dir = 'tb\sample')

NEURAL NETWORK PART: (A feed forward NN)

In [8]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(140,activation = 'relu'))
model.add(tf.keras.layers.Dense(220,activation = 'relu'))
model.add(tf.keras.layers.Dense(10,activation = 'softmax'))

Select your loss function and the type of optimizer to minimize it.(check keras website various losses and optimizers)

In [9]:
model.compile(optimizer = 'Adagrad',loss = 'sparse_categorical_crossentropy', metrics =['accuracy'])

Fit the flattened data.

 - Set callbacks to tensorboard object tb we created.

In [10]:
model.fit(flat_tr_x,tr_y,epochs = 5,verbose = 1,callbacks=[tb])

Epoch 1/5
60000/60000 [==============================] - 5s 85us/step - loss: 0.2054 - acc: 0.9386
Epoch 2/5
60000/60000 [==============================] - 5s 81us/step - loss: 0.1010 - acc: 0.9702
Epoch 3/5
60000/60000 [==============================] - 5s 79us/step - loss: 0.0764 - acc: 0.9775
Epoch 4/5
60000/60000 [==============================] - 5s 76us/step - loss: 0.0620 - acc: 0.9822
Epoch 5/5
60000/60000 [==============================] - 4s 73us/step - loss: 0.0526 - acc: 0.9850


You could see, loss is decreasing gradually and the accuracy is increasing.

# To view
- Open cmd, go the directory where the tb is created.
- Type 'tensorboard --logdir = tb\demo
- It will show you the address to go for tensorboard, copy that and paste in ur browser.

In [11]:
print(model.evaluate(flat_ts_x,ts_y))

10000/10000 [==============================] - 0s 35us/step
[0.07388780585532077, 0.9757]


The Testing accuracy is found to be 0.97. Dont be happy, it is pretty common with MNIST data.
CNN can achieve about 99% accuracy. 
Check my code on CNN.

The tensorflow tutorial of the same process is also available in my github check it.

Inspired from Sentdex.

# This is shortest NN code (Quite basic code) for beginners.
